sdk/cosmos/src/authorization_policy.rs

```diff
@@ -0,0 +1,320 @@
+                    let ret = &p[0..len - end_len - 1];
+                    return ret;
+                }
+            }
+        }
+    }
+    p
+}
+
+// TODO: make it private after pipeline migration
+pub(crate) fn generate_authorization(
+    auth_token: &AuthorizationToken,
+    http_method: &http::Method,
+    resource_type: &ResourceType,
+    resource_link: &str,
+    time: &str,
+) -> String {
+    let string_to_sign = string_to_sign(http_method, resource_type, resource_link, time);
+    debug!(
+        "generate_authorization::string_to_sign == {:?}",
+        string_to_sign
+    );
+
+    let str_unencoded = format!(
+        "type={}&ver={}&sig={}",
+        match auth_token {
+            AuthorizationToken::Primary(_) => "master",
+            AuthorizationToken::Resource(_) => "resource",
+        },
+        VERSION,
+        match auth_token {
+            AuthorizationToken::Primary(key) =>
+                Cow::Owned(encode_str_to_sign(&string_to_sign, key)),
+            AuthorizationToken::Resource(key) => Cow::Borrowed(key),
+        },
+    );
+    debug!(
+        "generate_authorization::str_unencoded == {:?}",
+        str_unencoded
+    );
+
+    form_urlencoded::byte_serialize(&str_unencoded.as_bytes()).collect::<String>()
+}
+
+fn string_to_sign(
+    http_method: &http::Method,
+    rt: &ResourceType,
+    resource_link: &str,
+    time: &str,
+) -> String {
+    // From official docs:
+    // StringToSign =
+    //      Verb.toLowerCase() + "\n" +
+    //      ResourceType.toLowerCase() + "\n" +
+    //      ResourceLink + "\n" +
+    //      Date.toLowerCase() + "\n" +
+    //      "" + "\n";
+    // Notice the empty string at the end so we need to add two new lines
+
+    format!(
+        "{}\n{}\n{}\n{}\n\n",
+        match *http_method {
+            http::Method::GET => "get",
+            http::Method::PUT => "put",
+            http::Method::POST => "post",
+            http::Method::DELETE => "delete",
+            http::Method::HEAD => "head",
+            http::Method::TRACE => "trace",
+            http::Method::OPTIONS => "options",
+            http::Method::CONNECT => "connect",
+            http::Method::PATCH => "patch",
+            _ => "extension",
+        },
+        match rt {
+            ResourceType::Databases => "dbs",
+            ResourceType::Collections => "colls",
+            ResourceType::Documents => "docs",
+            ResourceType::StoredProcedures => "sprocs",
+            ResourceType::Users => "users",
+            ResourceType::Permissions => "permissions",
+            ResourceType::Attachments => "attachments",
+            ResourceType::PartitionKeyRanges => "pkranges",
+            ResourceType::UserDefinedFunctions => "udfs",
+            ResourceType::Triggers => "triggers",
+        },
+        resource_link,
+        time.to_lowercase()
+    )
+}
+
+fn encode_str_to_sign(str_to_sign: &str, key: &[u8]) -> String {
```

Comment:
Even though all these functions are private or will become private, I think they could all use more documentation. How auth works in Azure isn't the most straightforward thing.